src/uu/echo/src/echo.rs

```diff
@@ -21,73 +21,98 @@ mod options {
+    // yield incorrect results because it will interpret values larger than
+    // `u8::MAX` as unicode.
+    let mut ret = input.peek().and_then(|c| c.to_digit(base as u32))? as u8;
+
+    // We can safely ignore the None case because we just peeked it.
+    let _ = input.next();
+
+    for _ in 1..base.max_digits() {
+        match input.peek().and_then(|c| c.to_digit(base as u32)) {
+            Some(n) => ret = ret.wrapping_mul(base as u8).wrapping_add(n as u8),
+            None => break,
+        }
+        // We can safely ignore the None case because we just peeked it.
+        let _ = input.next();
+    }
 
-    let mut buffer = ['\\'; 2];
+    Some(ret.into())
+}
 
-    // TODO `cargo +nightly clippy` complains that `.peek()` is never
-    // called on `iter`. However, `peek()` is called inside the
-    // `parse_code()` function that borrows `iter`.
+fn print_escaped(input: &str, mut output: impl Write) -> io::Result<bool> {
     let mut iter = input.chars().peekable();
-    while let Some(mut c) = iter.next() {
-        let mut start = 1;
-
-        if c == '\\' {
-            if let Some(next) = iter.next() {
-                c = match next {
-                    '\\' => '\\',
-                    'a' => '\x07',
-                    'b' => '\x08',
-                    'c' => {
-                        should_stop = true;
-                        break;
-                    }
-                    'e' => '\x1b',
-                    'f' => '\x0c',
-                    'n' => '\n',
-                    'r' => '\r',
-                    't' => '\t',
-                    'v' => '\x0b',
-                    'x' => parse_code(&mut iter, 16, 2, 4).unwrap_or_else(|| {
-                        start = 0;
-                        next
-                    }),
-                    '0' => parse_code(&mut iter, 8, 3, 3).unwrap_or('\0'),
-                    _ => {
-                        start = 0;
-                        next
-                    }
-                };
-            }
+    while let Some(c) = iter.next() {
+        if c != '\\' {
+            write!(output, "{c}")?;
+            continue;
         }
 
-        buffer[1] = c;
+        // This is for the \NNN syntax for octal sequences.
+        // Note that '0' is intentionally omitted because that
+        // would be the \0NNN syntax.
+        if let Some('1'..='8') = iter.peek() {
+            if let Some(parsed) = parse_code(&mut iter, Base::Oct) {
+                write!(output, "{parsed}")?;
+                continue;
+            }
+        }
 
-        // because printing char slices is apparently not available in the standard library
-        for ch in &buffer[start..] {
-            write!(output, "{ch}")?;
+        if let Some(next) = iter.next() {
+            let unescaped = match next {
+                '\\' => '\\',
+                'a' => '\x07',
+                'b' => '\x08',
+                'c' => return Ok(true),
```

Comment:
It is unclear what `true` means in this context. Maybe using a constant would be clearer? Or using an enum?

Reply:
I'll remind you that this PR is about fixing the wrapping behaviour, but sure 😄

What about `io::Result<ControlFlow<()>>`? Then `Ok(true)` becomes `Ok(Break(())`, `Ok(false)` becomes `Ok(Continue(()))` and `Err` stays the same.

Reply:
Looks good! I didn't know there is such an enum in the std, something learned :)